### PR TITLE
[V1][Core] Support offloading KV cache to CPU.

### PR DESCRIPTION
**Pull request by mengzhu28**
## TL;DR  
In V1, swap GPU KV cache blocks to CPU upon eviction and swap them back if there's a cache hit.  

## Swap Strategy  

CPU → GPU swap-in happens naturally when requests hit the cache (unless we do prefetching).  
GPU → CPU swap-out can be handled in two ways:  

1. **Eagerly**: Immediately after a request completes and its blocks are freed.  
2. **Lazily**: When evicting a GPU block while scheduling new requests.  

This PR adopts **(2)** to minimize unnecessary swaps. However, the downside is that the swap-out overhead might be exposed.  

Ideally, an optimal approach would asynchronously offload X cache blocks at a certain cadence (e.g., hidden behind the main CUDA graph) while maintaining free GPU block headroom. This would add complexity and is left for future work.  

## Implementation  

This PR builds on the excellent V1 KV cache manager, blend in with the existing interface.  
Newly introduced metadata states:  
- `cpu_block_pool` and `cached_block_hash_to_cpu_block` mirror their GPU counterparts.  

### High-Level Flow:  
- The KV cache manager accumulates swap-in/out decisions during each scheduling cycle.  
- These swap decisions are then "flushed" to the scheduler output, allowing model runners to issue aggregated swap calls before model execution, minimizing dispatch overhead.  

For simplicity, we avoid threading the scheduler output through multiple KV cache manager calls. Instead, swap-related data is accumulated in `step_*` fields (e.g., `step_h2d_swap_map`).  
A new `end_schedule_step` callback resets them at the end of each scheduling iteration. (Open to alternative designs.)

### CPU Cache Eviction Policy
We currently adopt a simple round-robin strategy to do CPU cache eviction. LRU will be added in a followup PR.

### User Configuration:  
We reuse the existing `--swap-space` flag (previously unused in V1) to control the number of CPU blocks.  
Whether to change the default (currently **4GB**) remains up for discussion.  

## Benchmark
TBA

## TODO
- [ ] write tests
- [ ] benchmarks and profiling
- [ ] docs

### CODE REVIEW

**Comment by BigCousin-z** [comment]
Is V0 Support？

**Comment by WoosukKwon** [comment]
Hi @mengzhu28, thanks for submitting the great PR! I will reach out to you offline.

File: vllm/v1/core/scheduler.py [hunk 1]
```diff
@@ -367,8 +377,12 @@ def schedule(self) -> "SchedulerOutput":
             # the previous and the current steps.
             finished_req_ids=self.finished_req_ids,
             free_encoder_input_ids=self.encoder_cache_manager.get_freed_ids(),
+            d2h_swap_map=self.kv_cache_manager.step_d2h_swap_map,
+            h2d_swap_map=self.kv_cache_manager.step_h2d_swap_map,
```
Comment (DearPlanet) [inline vllm/v1/core/scheduler.py:381]: Deep copy should be used here for `self.kv_cache_manager.step_d2h_swap_map` and self.kv_cache_manager.step_h2d_swap_map, otherwise their contents will be cleared in the next `self.kv_cache_manager.end_schedule_step()` and no blocks will be swapped.

Comment (DearPlanet) [reply vllm/v1/core/scheduler.py:381]: Besides this issue,

Comment (DearPlanet) [reply vllm/v1/core/scheduler.py:381]: Besides this issue, LGTM. This PR significantly enhances performance in use cases with large amount of prefix cache requirements. Great job!

File: vllm/v1/core/kv_cache_manager.py [hunk 2]
```diff
@@ -89,6 +101,16 @@ def __init__(
         self.num_cached_block: Dict[str, int] = defaultdict(int)
         self.prefix_cache_stats = PrefixCacheStats()
 
+        # The following swap maps are accumulated over a scheduling step.
+        # Then they are "flushed" as part of the scheduler output.
+        # GPU block ID -> CPU block ID
+        self.step_d2h_swap_map: Dict[int, int] = {}
```
Comment (maobaolong) [inline vllm/v1/core/kv_cache_manager.py:107]: Could you please do an abstract to support `Offload to Disk` in the future? If we did this abstraction, the data structure can be [ [src_device, dst_device] -> swap_map[src_block_id -> dst_block_id] ]. Any throughs?

**Review by maobaolong — COMMENTED** [review]
@mengzhu28 Thanks for this great work on V1, `offloading KV cache to CPU` can gains performance to TTFT and Throughput, just thinking about the next further step base on this PR, may be vllm can support `offloading KV cache to Disk` as followup work?

I left a comment inline about abstraction, please take a look, thanks.

File: tests/v1/test_utils.py [hunk 3]
```diff
@@ -23,7 +23,7 @@ def test_bind_kv_cache():
         'layers.3.self_attn': torch.zeros((1, )),
     }
     runner_kv_caches: List[torch.Tensor] = []
-    bind_kv_cache(kv_cache, ctx, runner_kv_caches)
+    bind_kv_cache(kv_cache, runner_kv_caches, ctx)
```
Comment (maobaolong) [inline tests/v1/test_utils.py:26]: Why you swap the two parameters?

File: vllm/v1/utils.py [hunk 4]
```diff
@@ -147,8 +147,10 @@ def shutdown(proc: multiprocessing.Process, input_path: str, output_path: str):
 
 def bind_kv_cache(
     kv_caches: Dict[str, torch.Tensor],
-    forward_context: Dict[str, "Attention"],
     runner_kv_caches: List[torch.Tensor],
+    forward_context: Dict[str, "Attention"],
```
Comment (maobaolong) [inline vllm/v1/utils.py:151]: change these parameters sequences can make more sense but in the other hand, it introduce more extra code changes.

File: vllm/v1/core/scheduler.py (hunk 1 above)
Comment (maobaolong) [reply vllm/v1/core/scheduler.py:381]: > end_schedule_step()

AFAIK, before next `self.kv_cache_manager.end_schedule_step()`, the `EngineCore#step` will call `model_executor.swap_blocks` and consume these two maps. Please correct me if i'm wrong.

image

Comment (maobaolong) [reply vllm/v1/core/scheduler.py:381]: @DearPlanet Good catch!

**Comment by WoosukKwon** [comment]
@mengzhu28 Could you please rebase the PR?

**Comment by mengzhu28** [comment]
@WoosukKwon as discussed offline, created RFC #16144

File: vllm/v1/core/scheduler.py (hunk 1 above)
Comment (chunxiaozheng) [reply vllm/v1/core/scheduler.py:381]: @mengzhu28 There is actually no swap out/in happens here, so the benchmark test results above only affect the speed, and there may be accuracy issues

**Comment by chunxiaozheng** [comment]
Would it be better to abstract the CPU offloading related functions into a new class and add a parameter to enable it?

File: vllm/v1/core/scheduler.py (hunk 1 above)
Comment (mengzhu28) [reply vllm/v1/core/scheduler.py:381]: good catch, thanks guys! I have pushed a quick fix to this branch. Will re-test it and address other comments.

File: vllm/v1/core/scheduler.py [hunk 5]
```diff
@@ -259,7 +262,10 @@ def schedule(self) -> "SchedulerOutput":
                     # improve the performance.
                     num_computed_tokens -= self.block_size
                     num_new_tokens = self.block_size
-                    computed_blocks.pop()
+                    if computed_blocks:
```
Comment (chunxiaozheng) [inline vllm/v1/core/scheduler.py:265]: The GPU hit must be before the CPU, so here we should first try `pop()` from the `computed_cpu_blocks `.

**Comment by jiawei-liang** [comment]
helo,can this support 1-cpu/n-gpu in one host situation?
